branchandbound.js

```diff
@@ -0,0 +1,143 @@
+var utils = require('./utils')
+
+var maxTries = 1000000
+
+function calculateEffectiveValues (utxos, feeRate) {
+  return utxos.map(function (utxo) {
+    if (isNaN(utils.uintOrNaN(utxo.value))) {
+      return {
+        utxo: utxo,
+        effectiveValue: 0
+      }
+    }
+
+    var effectiveFee = utils.inputBytes(utxo) * feeRate
+    var effectiveValue = utxo.value - effectiveFee
+    return {
+      utxo: utxo,
+      effectiveValue: effectiveValue
+    }
+  })
+}
+
+module.exports = function branchAndBound (utxos, outputs, feeRate) {
+  if (!isFinite(utils.uintOrNaN(feeRate))) return {}
+
+  var costPerOutput = utils.outputBytes({}) * feeRate
+  var costPerInput = utils.inputBytes({}) * feeRate
+  var costOfChange = costPerInput + costPerOutput
+
+  var outAccum = utils.sumOrNaN(outputs)
+
+  var effectiveUtxos = calculateEffectiveValues(utxos, feeRate).filter(function (x) {
+    return x.effectiveValue > 0
+  }).sort(function (a, b) {
+    return b.effectiveValue - a.effectiveValue
+  })
+
+  var selected = search(effectiveUtxos, outAccum, costOfChange)
+  if (selected != null) {
+    var inputs = []
+
+    for (var i = 0; i < effectiveUtxos.length; i++) {
+      if (selected[i]) {
+        inputs.push(effectiveUtxos[i].utxo)
+      }
+    }
+
+    return utils.finalize(inputs, outputs, feeRate)
+  } else {
+    const fee = feeRate * utxos.reduce(function (a, x) {
```

Comment:
It's a bug in the c++ code, now corrected. Anyway though, I noticed something weird.

When I corrected the error and add the overhead, the algorithm performs slightly *worse*. When I run, on the same randomly generated data, both the BnB with incorrect target and BnB with correct target, the BnB with incorrect targets always has better results. The difference is small, but consistent.

I have no idea why.

I will anyway add some tests finally, maybe there is some other subtle bug that I haven't found.